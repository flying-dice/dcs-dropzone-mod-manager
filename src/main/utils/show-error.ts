import { app, dialog, shell } from 'electron'
import { Logger } from '@nestjs/common'
import { config } from '../config'

const issueBodyTemplate = (error: Error, recentLogs: string[]) => `\

This is an automated error report generated by the application.

- **Error:** ${error.name}
- **Message:** ${error.message}
- **Stack Trace:**
\`\`\`text
${error.stack}
\`\`\`
---
[Please describe what you were doing when the error occurred here.]


Please Upload the logs from the following location:
${config.logfile.replace(app.getPath('appData'), '%APPDATA%')}

---
${recentLogs.join('\n')}
---

`

const title = 'Oops! Something Went Wrong'
const message = `\
An unexpected error occurred. Please try the following:

- Restart the application.
- Check your internet connection.
- Raise an issue if the problem persists.

An error report has been prepared for you and opened in the browser to share with the developers, should you choose to do so.
`

export function showError(error: Error, recentLogs: string[]) {
  const issueTitle = encodeURIComponent(`Error: ${new Date().toISOString()} Fatal Error Report`)

  const issueBody = encodeURIComponent(issueBodyTemplate(error, recentLogs))

  dialog.showErrorBox(title, message)

  shell.openExternal(
    `https://github.com/flying-dice/dcs-dropzone-mod-manager/issues/new?labels=bug&title=${issueTitle}&body=${issueBody}`
  )

  Logger.error(error, 'main')
}
