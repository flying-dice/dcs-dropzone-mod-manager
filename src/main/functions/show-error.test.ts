import { showError } from './show-error'
import { dialog, shell } from 'electron'
import { Logger } from '@nestjs/common'
import { vi } from 'vitest'

vi.mock('electron', () => ({
  dialog: { showErrorBox: vi.fn() },
  shell: { openExternal: vi.fn() }
}))

vi.mock('@nestjs/common', () => ({
  Logger: { error: vi.fn() }
}))

vi.mock('../logging', () => ({
  filename: 'test.log'
}))

describe('showError', () => {
  const error = new Error('Test error')
  error.stack = 'Error: Test error\n    at testFunction (test.js:1:1)'

  const recentLogs = ['log1', 'log2']

  afterEach(() => {
    vi.clearAllMocks()
  })

  it('displays an error dialog with the correct title and message', () => {
    showError(error, recentLogs)
    expect(dialog.showErrorBox).toHaveBeenCalledWith(
      'Oops! Something Went Wrong',
      expect.stringContaining('An unexpected error occurred.')
    )
  })

  it('should generate an error report matching snapshot', () => {
    showError(error, recentLogs)
    expect(shell.openExternal).toHaveBeenCalled()
    expect(Logger.error).toHaveBeenCalled()
    const issueBody = decodeURIComponent(
      vi.mocked(shell.openExternal).mock.calls[0][0].split('body=')[1]
    )
    expect(issueBody).toMatchSnapshot()

    const dialogBoxContent = vi.mocked(dialog.showErrorBox).mock.calls[0][1]
    expect(dialogBoxContent).toMatchSnapshot()
  })

  it('opens the GitHub issue URL with the correct parameters', () => {
    showError(error, recentLogs)
    expect(shell.openExternal).toHaveBeenCalledWith(
      expect.stringContaining('https://github.com/flying-dice/dcs-dropzone-mod-manager/issues/new')
    )
  })

  it('logs the error using the Logger', () => {
    showError(error, recentLogs)
    expect(Logger.error).toHaveBeenCalledWith(error, 'main')
  })

  it('handles empty recentLogs array', () => {
    showError(error, [])
    expect(dialog.showErrorBox).toHaveBeenCalled()
    expect(shell.openExternal).toHaveBeenCalled()
    expect(Logger.error).toHaveBeenCalled()
  })

  it('handles error with no stack trace', () => {
    const errorWithoutStack = new Error('Test error')
    errorWithoutStack.stack = ''
    showError(errorWithoutStack, recentLogs)
    expect(dialog.showErrorBox).toHaveBeenCalled()
    expect(shell.openExternal).toHaveBeenCalled()
    expect(Logger.error).toHaveBeenCalled()
  })
})
