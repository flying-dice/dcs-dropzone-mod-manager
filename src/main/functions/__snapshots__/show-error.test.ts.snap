// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`showError > should generate an error report matching snapshot 1`] = `
"
This is an automated error report generated by the application.

- **Error:** Error
- **Message:** Test error
- **Stack Trace:**
\`\`\`text
Error: Test error
    at testFunction (test.js:1:1)
\`\`\`
---
[Please describe what you were doing when the error occurred here.]


Please Upload the logs from the following location:
test.log

---
log1
log2
---

"
`;

exports[`showError > should generate an error report matching snapshot 2`] = `
"An unexpected error occurred. Please try the following:

- Restart the application.
- Check your internet connection.
- Raise an issue if the problem persists.

An error report has been prepared for you and opened in the browser to share with the developers, should you choose to do so.
"
`;
